core/src/main/java/io/grpc/internal/TransportTracer.java

```diff
@@ -0,0 +1,212 @@
+   * window size.
+   */
+  public void setLocalFlowControlWindowPollable(Callable<Integer> localFlowControllPollable) {
+    this.localFlowControlPollable = localFlowControllPollable;
+  }
+
+  /**
+   * Returns the number of streams started on the transport.
+   */
+  public long getStreamsStarted() {
+    return streamsStarted;
+  }
+
+  /**
+   * Returns the number of streams ended successfully with an OK status.
+   */
+  public long getStreamsSucceeded() {
+    return streamsSucceeded.get();
+  }
+
+  /**
+   * Returns the number of streams completed with a non-OK status.
+   */
+  public long getStreamsFailed() {
+    return streamsFailed.get();
+  }
+
+  /**
+   * Returns the number of messages sent on the transport.
+   */
+  public long getMessagesSent() {
+    return messagesSent.get();
+  }
+
+  /**
+   * Returns the number of messages received on the transport.
+   */
+  public long getMessagesReceived() {
+    return messagesReceived.get();
+  }
+
+  /**
+   * Returns the number of keep alive messages sent on the transport.
+   */
+  public long getKeepAlivesSent() {
+    return keepAlivesSent.get();
+  }
+
+  /**
+   * Returns the last time a stream was created as millis since Unix epoch.
+   */
+  public long getLastStreamCreatedTimeMsec() {
+    return lastStreamCreatedTimeMsec;
+  }
+
+  /**
+   * Returns the last time a message was sent as millis since Unix epoch.
+   */
+  public long getLastMessageSentTimeMsec() {
+    return lastMessageSentTimeMsec.get();
+  }
+
+  /**
+   * Returns the last time a message was received as millis since Unix epoch.
+   */
+  public long getLastMessageReceivedTimeMsec() {
+    return lastMessageReceivedTimeMsec.get();
+  }
+
+  /**
+   * Returns the remote flow control window as reported by the callback of
+   * {@link #setRemoteFlowControlWindowPollable}. Returns -1 if no callback was registered.
+   */
+  public int getRemoteFlowControlWindow() {
+    if (remoteFlowControlPollable == null) {
+      return -1;
+    }
+    try {
+      return remoteFlowControlPollable.call();
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  /**
+   * Returns the local flow control window as reported by the callback of
+   * {@link #setRemoteFlowControlWindowPollable}. Returns -1 if no callback was registered.
+   */
+  public int getLocalFlowControlWindow() {
+    if (localFlowControlPollable == null) {
+      return -1;
+    }
+    try {
+      return localFlowControlPollable.call();
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  /**
+   * Updates an AtomicLong representing a msec timestamp. Avoids races and only allows the value
+   * to increase.
+   */
+  private static void updateMsecTimestamp(AtomicLong timestamp) {
+    long now = System.currentTimeMillis();
+    long oldVal = timestamp.get();
+    while (oldVal < now && !timestamp.compareAndSet(oldVal, now)) {
```

Comment:
A second alternative would be getAndSet.  Most of the time, the clock goes forward.  getAndSet would allow you to check if the most recent time was actually  greater than timestamp, and let you compareAndSet to roll it back.

That would avoid the initial read too.